### PR TITLE
Sidecar/Ruler/Receiver: Add flag to disable upload of debug meta files

### PR DESCRIPTION
**Pull request by hitanshu-mehta**
Signed-off-by: Hitanshu Mehta <hitanshu99amehta@gmail.com>

Fixes: #3839 

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes
- Introduced new flag `--shipper.upload-debug-meta-files` in sidecar, ruler and receiver.
- Added test to verify working of flag.

## Verification
- Ran unit and e2e tests locally.

### CODE REVIEW

File: pkg/block/block.go
Comment (bwplotka) [inline pkg/block/block.go:106]: Hm, decision was to remove this.. optionally? 

TBH I never used, but wonder if this was not used by someone at the end.

Comment (bwplotka) [reply pkg/block/block.go:106]: One way would be to remove .. let's say by default but enable for opt-in? 🤔

**Review by bwplotka — APPROVED** [review]
Nice. Let's decide if we want to remove completely or enable optionally.

File: pkg/block/block.go
Comment (pracucci) [reply pkg/block/block.go:106]: I'm fine to fully remove it. I know @pstibrany used it few times, so let me ask it too.

Comment (pstibrany) [reply pkg/block/block.go:106]: I can only repeat what I wrote here: https://github.com/thanos-io/thanos/issues/3839#issuecomment-786674853

I don’t have strong feelings about it, but found it useful in the past.

Comment (bwplotka) [reply pkg/block/block.go:106]: Me too, that's why I created this in first place

**Comment by bwplotka** [comment]
Can we move this under some flag for now? 🤗

**Comment by hitanshu-mehta** [comment]
Sure. Flag named `--shipper.upload-debug-meta-files` in sidecar seems correct?

**Comment by yeya24** [comment]
> Sure. Flag named `--shipper.upload-debug-meta-files` in sidecar seems correct?

Receiver and Ruler support uploading blocks as well.

**Comment by hitanshu-mehta** [comment]
Thank you. I will add flags in them too.

File: cmd/thanos/downsample.go
Comment (bwplotka) [inline cmd/thanos/downsample.go:319]:
```suggestion
	if err := block.Upload(ctx, logger, bkt, resdir, hashFunc, true); err !=nil
```

File: pkg/block/block.go
Comment (bwplotka) [inline pkg/block/block.go:101]: I think we need uploader struct at this point. Will add this in another PR - too many args TBH (:

**Review by bwplotka — COMMENTED** [review]
Nice, some suggestion and idea for next step (:

File: pkg/block/block.go
Comment (hitanshu-mehta) [reply pkg/block/block.go:101]: Sure 😊

File: cmd/thanos/tools_bucket.go
Comment (prmsrswt) [inline cmd/thanos/tools_bucket.go:910]: The same flag makes sense here too.

File: pkg/block/block_test.go
Comment (prmsrswt) [inline pkg/block/block_test.go:214]: Do we have a test that checks the `Upload block with debug meta files.` condition as well?

File: pkg/verifier/index_issue.go
Comment (prmsrswt) [inline pkg/verifier/index_issue.go:114]: This needs to passed through a flag as well I suppose.

File: pkg/verifier/safe_delete.go
Comment (prmsrswt) [inline pkg/verifier/safe_delete.go:139]: Ditto, we need a flag here too.

File: cmd/thanos/config.go
Comment (prmsrswt) [inline cmd/thanos/config.go:138]: We are using the value `true` by default in a lot of places (tests etc) so why the default is "false"? IMO it makes more sense to have the default as "true" as it will maintain backward compatibility as well.

File: pkg/block/block_test.go
Comment (hitanshu-mehta) [reply pkg/block/block_test.go:214]: Yes. There is one test case above this. https://github.com/thanos-io/thanos/blob/816628343d4de1a9b8f829178906b3387addd88c/pkg/block/block_test.go#L139

Comment (prmsrswt) [reply pkg/block/block_test.go:214]: That test is not verifying the presence of debug meta files, it should. Also, this test that you added, you should check that the debug meta files are NOT uploaded, you are not checking that.

Comment (hitanshu-mehta) [reply pkg/block/block_test.go:214]: If we upload with debug meta files then 4 files are uploaded per block. So the test I wrote there are only 7 files in the obj. store (4 files from the test with debug meta files + 3 files from the test without debug metafiles).

Comment (prmsrswt) [reply pkg/block/block_test.go:214]: Gotcha :+1:  But I think, like we are verifying ChunkFiles, Meta File etc, we can explicitly verify debug metas too.

Comment (hitanshu-mehta) [reply pkg/block/block_test.go:214]: Understood :) I will add that verification. 
I have 2 questions.
1. Should I add that verification in all test cases?
2. Is it good practice that one test case depends on another test case? Because the test case I wrote depends on the previous one. I decided to write this test case this way because otherwise, I have to make a new instance obj. store. WDYT? 🤗

Comment (prmsrswt) [reply pkg/block/block_test.go:214]: 1. No need to add it in other tests, it'd be redundant.
2. In this case, I think it's not a good practice. What if someone adds a new test case in between?

Comment (hitanshu-mehta) [reply pkg/block/block_test.go:214]: Got it 👍🏻 . I will change the test case accordingly. Thank you :)

**Comment by hitanshu-mehta** [comment]
PR is ready for review 🤗

**Review by bwplotka — DISMISSED** [review]
Looks good! Just rebase. I hope we can improve the `Upload` someday with optional attributes./parameters.

**Comment by hitanshu-mehta** [comment]
Thank you for the review :) I will create a follow-up PR for uploader struct.

**Comment by bwplotka** [comment]
We can see that we touch so many things...

Given you need to rebase any way, maybe it's worth creating uploader struct... in this PR? (:

**Comment by hitanshu-mehta** [comment]
Yes, I agree. To rebase, I have to resolve many merge conflicts.

**Comment by pstibrany** [comment]
If keeping support for use in Cortex was the only reason to keep the flag, I would suggest to simplify the PR, drop the flag and remove support for those debug meta files completely.

We can already add support for debug/meta.json file on our own easily, if we really want to. (eg. by having a special bucket that uploads meta.json to two places).

**Comment by kakkoyun** [comment]
Hey @hitanshu-mehta. We're going to cut a release candidate soon, is there any chance for you to push this to the finish line?

**Comment by hitanshu-mehta** [comment]
Yes, I will be able to do it. I was waiting for the decision on whether we going to completely remove support for debug meta files or not 😅

**Comment by hitanshu-mehta** [comment]
I guess, failed e2e test is not related.

**Comment by kakkoyun** [comment]
> Looks like we decided to have this opt-in by default ([#3852 (comment)](https://github.com/thanos-io/thanos/pull/3852#discussion_r584614558)). Then maybe let's add an entry in the CHANGELOG.

I think the only decision to make is to enable this by default or not? @bwplotka @onprem @pracucci @pstibrany

**Comment by yeya24** [comment]
Let's revisit this. My opinion now is to remove debug metas totally and we don't need a flag.

**Comment by yeya24** [comment]
Let's get back to this finally. @hitanshu-mehta Are you able to continue this? Thank you!

**Comment by hitanshu-mehta** [comment]
@yeya24 Yes. So the final decision is to remove debug meta files completely without any flag. right?

**Comment by yeya24** [comment]
> @yeya24 Yes. So the final decision is to remove debug meta files completely without any flag. right?

Yes! Thanks for the patience.

**Comment by hitanshu-mehta** [comment]
I apologize for the delay from my side. @r0mdau has opened a PR #5110 fixing this issue. Hence, we can close this PR.

**Pull request closed without merging**
